### PR TITLE
Add support to DAC for reading tagged memory on tracked types

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Objective-C interop was added to CoreCLR in [dotnet/runtime#52146](https://github.com/dotnet/runtime/pull/52146). The [design document](https://github.com/dotnet/designs/blob/main/accepted/2021/objectivec-interop.md) for this feature included support for new [diagnostic features](https://github.com/dotnet/designs/blob/main/accepted/2021/objectivec-interop.md#diagnostics) in the [SOS tooling](https://github.com/dotnet/diagnostics). This document attempts to capture those details.

The following capability should be provided by SOS. All mentioned .NET types are contained within the `System.Runtime.InteropServices.ObjectiveC` namespace unless stated otherwise.

The `ObjectiveCMarshal.CreateReferenceTrackingHandle()` API allocates a `GCHandle` of type [`HNDTYPE_REFCOUNTED`](https://github.com/dotnet/runtime/blob/8818c1561089dfc4a7b4296982ee2f82a30c3c1d/src/coreclr/gc/gcinterface.h#L377-L387) for the passed in `object` instance. This handle type should already be known to SOS as it is also used for COM interop. What is new to this usage is "tagged memory". The tagged memory is presently 2 pointer sizes worth (that is, `sizeof(void*) * 2`) but could expand in the future. This memory is stored on the `InteropSyncBlockInfo` type in the [`m_taggedMemory`](https://github.com/dotnet/runtime/blob/8818c1561089dfc4a7b4296982ee2f82a30c3c1d/src/coreclr/vm/syncblk.h#L914) field. 

**SOS**: When examining an `object` (that is, `!do`) and/or an `object`'s `SyncBlock` (that is, `!SyncBlk`) provide the address of the tagged memory and size.

Once this work is completed the design document's [Diagnostics section](https://github.com/dotnet/designs/blob/main/accepted/2021/objectivec-interop.md#diagnostics) will be updated to reflect the commands.

### CODE REVIEW

File: src/coreclr/vm/syncblk.h [hunk 1]
```diff
@@ -894,7 +894,7 @@ class InteropSyncBlockInfo
         LIMITED_METHOD_CONTRACT;
```
Comment (hoyosjs) [inline src/coreclr/vm/syncblk.h:894]: I am not sure `AllocTaggedMemory` should be in the DAC as we don't tend to have mutating APIs in the DAC. 

Also, I `GetTaggedMemory`,  `m_taggedMemory`, and `m_taggedAlloc` are not properly DACized and could be dangerous. The main problem is if this API gets somehow called, and someone tried to read the pointer, they'll read in the address space of the debugger. I see that this is supposed to be "scratch memory", so maybe it doesn't make sense to read into it, but given they are pointers. If it always just reads from `m_taggedAlloc` and having a copy on the debugger side and trying to remote read this through the DAC makes no sense, then sure. I just have a knee jerk reaction over undacized pointers being "we won't need this" landmines for people working in diagnostics.

**Review by hoyosjs — APPROVED** [review]

File: src/coreclr/vm/syncblk.h (hunk 1 above)
Comment (noahfalk) [reply src/coreclr/vm/syncblk.h:894]: I'd suggest:
void* m_taggedMemory -> PTR_VOID m_taggedMemory
void* GetTaggedMemory() -> PTR_VOID GetTaggedMemory()

In a non-DAC build PTR_VOID is defined as void* so product build is unchanged. When compiled in the DAC PTR_VOID is a smart pointer that unambiguously refers to the debuggee address space. If you attempt to read memory through it (which this current change never does) the smart pointer would copy the appropriate memory from the debuggee.

The risk entailed by having AllocTaggedMemory() compiled into the DAC build pre-dated this change, but while we are staring at it and talking about it we may as well tidy it up. Wrapping #ifndef DACCESS_COMPILE/#endif around the method eliminates the possibility of DAC code ever invoking it.

**Review by noahfalk — APPROVED** [review]
LGTM aside from the already commented stuff. Thanks @AaronRobinsonMSFT !

File: src/coreclr/vm/syncblk.h (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/syncblk.h:894]: Thanks for the guidance and explanation of concern.

File: src/coreclr/debug/daccess/request.cpp [hunk 2]
```diff
@@ -5040,3 +5040,99 @@ HRESULT ClrDataAccess::GetComWrappersRCWData(CLRDATA_ADDRESS rcw, CLRDATA_ADDRES
...
+        return E_INVALIDARG;
+    }
+
+    *taggedMemory = NULL;
+    *taggedMemorySizeInBytes = 0;
+
+    SOSDacEnter();
+
+#ifdef FEATURE_OBJCMARSHAL
+    EX_TRY_ALLOW_DATATARGET_MISSING_MEMORY
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/debug/daccess/request.cpp]: Done

**Review by hoyosjs — APPROVED** [review]

**Pull request merged**